### PR TITLE
Add stop functionality for output config changes

### PR DESCRIPTION
**Pull request by fearful-symmetry**
## What does this PR do?
Fixes https://github.com/elastic/elastic-agent/issues/1913

This adds a reloader component, set by a `output_restart` config flag, that will stop the beat when it receives a config change to the output. 

This will require some other component changes before it works properly

- The agent needs to understand that the beat will restart on output change. Currently this will just result in the beats being marked as `FAILED`.
- For beats where we want this enabled, we need to add `-E management.output_restart=true` to the beat specfile. 

Testing this is easy. Just add the aforementioned line to the specfile, run the beat, then change the config. If there's a change to the output, the beats will shut down, agent will restart with the updated config.

## Why is it important?

We still want output changes to force a reload of the beat.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### CODE REVIEW

**Comment by cmacknz** [comment]
> The agent needs to understand that the beat will restart on output change. Currently this will just result in the beats being marked as FAILED.

Does it eventually recover from this? Or does it just stay in the failed state?

File: x-pack/libbeat/management/managerV2.go [hunk 1]
```diff
@@ -359,6 +377,8 @@ func (cm *BeatV2Manager) handleOutputReload(unit *client.Unit) {
 		_ = unit.UpdateState(client.UnitStateFailed, errString.Error(), nil)
 		return
 	}
+	// set to true, we'll reload the output if we need to re-configure
+	cm.outputIsConfigured = true
```
Comment (cmacknz) [inline x-pack/libbeat/management/managerV2.go:381]: This is here as a guard against restarting the first time the output is ever configured?

File: x-pack/libbeat/management/managerV2.go [hunk 2]
```diff
@@ -338,6 +346,16 @@ func (cm *BeatV2Manager) handleOutputReload(unit *client.Unit) {
 	}
 	cm.logger.Debugf("Got Output unit config '%s'", rawConfig.GetId())
 
+	// if needed, stop the beat, let agent restart
```
Comment (cmacknz) [inline x-pack/libbeat/management/managerV2.go:349]: We are sure this only triggers when the actual output configuration changes? And it won't restart for changes in the unit that are not the actual output configuration, like the log level? https://github.com/elastic/elastic-agent-client/blob/15881a8e64ef95f32f688c933319cd926ff500b0/elastic-agent-client.proto#L280

File: x-pack/libbeat/management/managerV2.go (hunk 1 above)
Comment (fearful-symmetry) [reply x-pack/libbeat/management/managerV2.go:381]: yup.

File: x-pack/libbeat/management/managerV2.go (hunk 2 above)
Comment (fearful-symmetry) [reply x-pack/libbeat/management/managerV2.go:349]: Ah, didn't think of that! Should probably add some kind of check to diff the actual config.

**Comment by fearful-symmetry** [comment]
for anyone reading: holding off on this until https://github.com/elastic/beats/pull/34066 is merged.

**Comment by cmacknz** [comment]
> The agent needs to understand that the beat will restart on output change. Currently this will just result in the beats being marked as FAILED.

> Does it eventually recover from this? Or does it just stay in the failed state?

Spoke with Alex about this. The FAILED state lasts a second or two while the agent detects the process is stopped and restarts. If this happens right before the next checkin with fleet server we may have the agent reported unhealthy unnecessarily, both in that the behaviour causing the failed state is intentional and that the status won't clear until the next checkin up to 5 minutes later.

To avoid this we need to debounce the process state locally as described in https://github.com/elastic/elastic-agent/issues/1946

**Comment by cmacknz** [comment]
These changes have been merged into https://github.com/elastic/beats/pull/34066, closing.

**Pull request closed without merging**